### ISSUE TITLE
Block Internet if in background

### ISSUE THREAD

**Issue by kasnder** [OxfordHCC/tracker-control-android#89]
This would be a really nice addition to TrackerControl

**Comment by kasnder** [comment]
This explains how to do this:

https://stackoverflow.com/questions/3873659/android-how-can-i-get-the-current-foreground-activity-from-a-service/27642535#27642535

**Comment by DrTexx** [comment]
+1 This would be absolutely excellent :pray:

My availability is hit and miss, however please let me know if there's any way I can assist with this outside of me requiring a proficiency in Java. I'd be more than happy to perform testing or give feedback on proposed/implemented UX :)

**Comment by kasnder** [comment]
Thanks for getting in touch. If you have any concrete ideas how to improve the app, I’d love to hear those ideas! :) Maybe, you already have some ideas what the UX could look like?

I’ve already thought out the technical solution, but haven’t implemented it yet.

**Comment by kasnder** [comment]
This is implemented in RethinkApp.

https://github.com/celzero/rethink-app/blob/a4364fd8213b9b4c9dc434c3c974401b802f1fa9/app/src/main/AndroidManifest.xml#L72

https://github.com/celzero/rethink-app/blob/a4364fd8213b9b4c9dc434c3c974401b802f1fa9/app/src/main/java/com/celzero/bravedns/util/BackgroundAccessibilityService.kt#L86

https://github.com/celzero/rethink-app/blob/a4364fd8213b9b4c9dc434c3c974401b802f1fa9/app/src/main/java/com/celzero/bravedns/automaton/FirewallManager.kt#L111

The implementation is still beta.

**Comment by kasnder** [comment]
There’s also this sample implementation: https://github.com/seguri/GetForegroundActivity

**Comment by ignoramous** [comment]
Hi @kasnder 

One of the engineers on RethinkDNS here..

We are also looking at the [UsageStatsManager APIs](https://developer.android.com/reference/android/app/usage/UsageStatsManager#queryEvents(long,%20long)) to see if there's a leaner way to implement this feature as using AccessibilityManager does tend to increase battery usage a bit (though we should filter for interesting events via AndroidManifest.xml to reduce that hit a bit).

**Comment by kasnder** [comment]
> We are also looking at the [`UsageStatsManager` APIs](https://developer.android.com/reference/android/app/usage/UsageStatsManager#queryEvents(long,%20long)) to see if there's a leaner way to implement this feature as using `AccessibilityManager` does tend to increase battery usage a bit (though we should filter for interesting events via `AndroidManifest.xml` to reduce that hit a bit).

Really great to hear from you! It seems both APIs come with flaws. I heard that `UsageStatsManager` comes with a certain delay, and not sure what the battery impact it has.

Finding a robust, root-free solution promises great benefits for users, so I’m very interested in investigating this further.

**Comment by kasnder** [comment]
One of my users suggested that the 'data saver' option rather reliably blocks network traffic in the background. Would be interesting to verify this claim.

**Comment by kasnder** [comment]
Another option might be to take inspiration from Greenify to freeze apps in the background. So, there's the question of how much an app like TC or RethinkDNS needs to offer.

**Comment by iiitomo** [comment]
hi, some apps need internet in background, e.g. for background music streaming. If this feature is enabled by default in TC, users would have hard time understanding why their music stream stops when app is in background, they would probably first think this is a bug in their music app.

And TrackerControl's name suggests its main goal is to block trackers, not blocking internet. Currently TC blocks trackers even for apps in background, correct ?

So I think this feature may be a plus for advanced users, but it should be **disabled by default** in TC, and when enabled the user needs a  'Allow internet when app in background' toggle **per app**.

**Comment by ignoramous** [comment]
> One of my users suggested that the 'data saver' option rather reliably blocks network traffic in the background. Would be interesting to verify this claim.

Yes, this works. Apps get sent the dreaded [`BLOCKED`](https://cs.android.com/android/platform/superproject/+/master:frameworks/base/services/core/java/com/android/server/ConnectivityService.java;l=1517;drc=ae67988721788b99b6150aeb8904c1c9d432fb0e) message when they try to access Internet when "data-saver" is enabled. On OnePlus, there's an option in Android Settings' app-info pages against every app to disable `Background data` which works in a similar manner too, but I don't think non-privileged apps have necessary permissions to flip [that setting programmatically](https://cs.android.com/android/platform/superproject/+/master:packages/apps/Settings/src/com/android/settings/datausage/DataSaverBackend.java;l=72;drc=0a476f5f533c316a47ab13cf22d7b12b96263117). That said, `AccessiblityService` APIs can be used to toggle [that preference on/off](https://cs.android.com/android/platform/superproject/+/master:packages/apps/Settings/src/com/android/settings/datausage/DataSaverSummary.java;l=99-107;drc=966fa01423b70424abbfc6c5029088a354e5d00c) (that is, perform UI actions on behalf of the user).

**Comment by kasnder** [comment]
> > One of my users suggested that the 'data saver' option rather reliably blocks network traffic in the background. Would be interesting to verify this claim.
> 
> Yes, this works. Apps get sent the dreaded [`BLOCKED`](https://cs.android.com/android/platform/superproject/+/master:frameworks/base/services/core/java/com/android/server/ConnectivityService.java;l=1517;drc=ae67988721788b99b6150aeb8904c1c9d432fb0e) message when they try to access Internet while in the background when "data-saver" is enabled. On OnePlus, there's an option in Android Settings' app-info pages against every app to disable `Background data` which works in a similar manner too, but I don't think non-privileged apps have [the necessary permissions](https://cs.android.com/android/platform/superproject/+/master:frameworks/base/services/core/java/com/android/server/net/NetworkPolicyManagerService.java;l=3050;drc=f072281f634e192638431bb4e93cb75fdf63d8da) to flip [that setting programmatically](https://cs.android.com/android/platform/superproject/+/master:packages/apps/Settings/src/com/android/settings/datausage/DataSaverBackend.java;l=72;drc=0a476f5f533c316a47ab13cf22d7b12b96263117). That said, `AccessiblityService` APIs can be used to toggle [that preference on/off](https://cs.android.com/android/platform/superproject/+/master:packages/apps/Settings/src/com/android/settings/datausage/DataSaverSummary.java;l=99-107;drc=966fa01423b70424abbfc6c5029088a354e5d00c) (that is, perform UI actions on behalf of the user).

Hi @ignoramous! Thank you so much for this update. Your implementation looks really clean. Well done! I'll definitely check it out, to potentially help TrackerControl users, too, to avoid unwanted tracking in the background.
